fix: superglobals-realated crash with custom extensions in worker mode

dunglas opened the pull request:
Closes #767.
Partial fix for #592 (memory leak when dealing with cookies in worker mode).

dunglas commented:
@withinboredom @TimWolla would you mind reviewing the proposed fix?

I need to do more testing to ensure that the behavior is what is expected, but at least this fixes the segfault. We'll have to double-check if this doesn't introduce a memory leak or some weird issues.

TimWolla commented on frankenphp.c:
```diff
@@ -69,6 +69,7 @@ typedef struct frankenphp_server_context {
   uintptr_t current_request;
   uintptr_t main_request;
   bool worker_ready;
+  zval **worker_http_globals;
```
```suggestion
  zval *worker_http_globals;
```

TimWolla commented on frankenphp.c:
```diff
@@ -105,13 +106,23 @@ static uintptr_t frankenphp_clean_server_context() {
+             sizeof(zval) * NUM_TRACK_VARS);
+      php_hash_environment();
+    } else {
+      ctx->worker_http_globals = (zval **)malloc(sizeof(zval) * NUM_TRACK_VARS);
```
Casting the return value of `malloc()` is not necessary in C. In this case it luckily pointed out the incorrect type of `worker_http_globals`, though.

TimWolla commented on frankenphp.c:
```diff
@@ -105,13 +106,23 @@ static uintptr_t frankenphp_clean_server_context() {
+      php_hash_environment();
+    } else {
+      ctx->worker_http_globals = (zval **)malloc(sizeof(zval) * NUM_TRACK_VARS);
+      memcpy(ctx->worker_http_globals, &PG(http_globals),
+             sizeof(zval) * NUM_TRACK_VARS);
```
And here it probably should be this (untested):
```suggestion
      memcpy(ctx->worker_http_globals, PG(http_globals),
             sizeof(zval) * NUM_TRACK_VARS);
```

TimWolla reviewed (commented):
> I need to do more testing to ensure that the behavior is what is expected, but at least this fixes the segfault. We'll have to double-check if this doesn't introduce a memory leak or some weird issues.

It does leak:

```
[Tue May 21 07:27:50 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fff802020a0 (40 bytes), script=-
Last leak repeated 14 times
[Tue May 21 07:27:50 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fff802020f0 (40 bytes), script=-
Last leak repeated 42 times
[Tue May 21 07:27:50 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fff802025a0 (40 bytes), script=-
Last leak repeated 12 times
[Tue May 21 07:27:50 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_hash.c(291) :  Freeing 0x00007fff8025f000 (56 bytes), script=-
Last leak repeated 4 times
[Tue May 21 07:27:50 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_hash.c(1311) :  Freeing 0x00007fff80260c00 (2560 bytes), script=-
=== Total 77 memory leaks detected ===
```

For debugging memory leaks, I suggest `USE_ZEND_ALLOC=0` + ASAN. Valgrind does not appear to work well here.

I'm also seeing 

```
2024/05/21 07:27:50.165 INFO    restarting      {"worker": "/go/src/app/testdata/worker.php"}
```

immediately after, but I'm not sure if that is expected.

The following test script also appears to log the incorrect information:

```
<?php

ignore_user_abort(true);

for($running = true; $running;) {
    file_put_contents(__DIR__ . '/log', "before: " . $_SERVER['REQUEST_URI'] . ' ' . $_SERVER['REMOTE_ADDR'] . "\n", FILE_APPEND);

    $running = \frankenphp_handle_request(static function () {
        file_put_contents(__DIR__ . '/log', "inside: " . $_SERVER['REQUEST_URI'] . ' ' . $_SERVER['REMOTE_ADDR'] . "\n", FILE_APPEND);
    });

    file_put_contents(__DIR__ . '/log', "after : " . $_SERVER['REQUEST_URI'] . ' ' . $_SERVER['REMOTE_ADDR'] . "\n", FILE_APPEND);
}
```

I would expected `before:` and `after:` to look identical at all times, but I'm seeing:

```
before: worker.php 
before: worker.php 
before: worker.php 
before: worker.php 
inside: /worker.php?some_query_parameters 127.0.0.1
after : /worker.php?some_query_parameters 127.0.0.1
before: /worker.php?some_query_parameters 127.0.0.1
inside: /worker.php?some_query_parameters 127.0.0.1
after : /worker.php?some_query_parameters 127.0.0.1
before: /worker.php?some_query_parameters 127.0.0.1
```

TimWolla commented:
> Valgrind does not appear to work well here.

I does point out some conditional jumps that depend on uninitialized in non-go code values, though. But also quite a few invalid reads in go code, which might or might not be expected.

TimWolla commented:
> The following test script also appears to log the incorrect information:

The cause probably is that you can't naively `memcpy` the zval, because the reference counting of strings is then incorrect, allowing the code to perform in-place modification instead of doing copy-on-write. I didn't test, because I find the DX of the development container to be pretty unwieldy.

dunglas commented:
Thanks for your detailed review @TimWolla. Regarding the Docker container, it is entirely optional. When you have a local installation of PHP, you can just run this command to compile the binary: https://frankenphp.dev/docs/compile/#compile-the-go-app

withinboredom commented on frankenphp.c:
```diff
@@ -105,13 +106,23 @@ static uintptr_t frankenphp_clean_server_context() {
 
-    memset(&PG(http_globals), 0, sizeof(zval) * NUM_TRACK_VARS);
+      // Restore worker script super globals
+      memcpy(&PG(http_globals), ctx->worker_http_globals,
```
Pretty sure this will break GC since reference counting won't be performed and may either cause a memory leak or GC'ing of variables unexpectedly.

What you probably (maybe) want to do is create a hashmap copy of globals before starting the work, then after the work, restore that hashmap to globals

dunglas commented:
I pushed a new version that uses `ZVAL_COPY` as suggested by @TimWolla. This should fix the leak (to be double-checked).

This behaves almost as expected but I identified two issues:

* [ ] `$_SERVER['FRANKENPHP_WORKER']` and probably other custom environment variables passed to the worker script are lost inside the closure. We should probably merge the variables from the worker script in the closure to prevent this issue.
* [ ] `$_SERVER['FRANKENPHP_WORKER']`  is NULL terminated, that's an existing, unrelated bug. I'll fix ti.

dunglas commented:
The behavior is now the expected one. If you don't mind, cloud you please review again?

dunglas commented:
Actually, merging the $_SERVER superglobals of the worker and the request wasn't expected, we even have a test (the failing one) for that.

As showcased by the test, it's entirely possible to do that userland if needed.

Should we keep the existing behavior of not merging (after a second thought, I think that it's more flexible to not merge)?

TimWolla commented:
I'm still seeing a leak:

```
b57e622520ad# ../caddy/frankenphp/frankenphp php-server --worker worker-getopt.php
2024/05/22 13:44:51.696 WARN    admin   admin endpoint disabled
2024/05/22 13:44:51.696 WARN    http.auto_https server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server {"server_name": "php", "http_port": 80}
2024/05/22 13:44:51.696 INFO    tls.cache.maintenance   started background certificate maintenance      {"cache": "0xc000345a80"}
2024/05/22 13:44:51.713 INFO    FrankenPHP started 🐘   {"php_version": "8.3.8-dev"}
2024/05/22 13:44:51.714 INFO    http.log        server running  {"name": "php", "protocols": ["h1", "h2", "h3"]}
2024/05/22 13:44:51.714 INFO    Caddy serving PHP app on :80
2024/05/22 13:44:51.723 WARN    tls     storage cleaning happened too recently; skipping for now        {"storage": "FileStorage:/root/.local/share/caddy", "instance": "00d35d99-1901-4bf0-aa0b-e1f662db2acd", "try_again": "2024/05/23 13:44:51.723", "try_again_in": 86399.999999265}
2024/05/22 13:44:51.723 INFO    tls     finished cleaning storage units
[Wed May 22 13:44:53 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fe08c2021e0 (40 bytes), script=-
Last leak repeated 12 times
[Wed May 22 13:44:53 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fe08c202230 (40 bytes), script=-
Last leak repeated 40 times
[Wed May 22 13:44:53 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_string.h(174) :  Freeing 0x00007fe08c2026e0 (48 bytes), script=-
Last leak repeated 12 times
[Wed May 22 13:44:53 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_hash.c(291) :  Freeing 0x00007fe08c25f000 (56 bytes), script=-
Last leak repeated 4 times
[Wed May 22 13:44:53 2024]  Script:  '-'
/usr/local/src/php/Zend/zend_hash.c(1311) :  Freeing 0x00007fe08c260c00 (2560 bytes), script=-
=== Total 73 memory leaks detected ===
2024/05/22 13:44:53.542 ERROR   unexpected termination, restarting      {"worker": "/go/src/app/testdata/worker-getopt.php", "exit_status": 2}
```

and then calling `curl -v http://localhost/worker-getopt.php?some_query_parameters`.

I also might've done something wrong while starting FrankenPHP, because of the `exit(2)` being triggered.

TimWolla commented:
And the test script with the `file_put_contents` from yesterday’s review still outputs the wrong data for the `after:`.

dunglas commented:
I've an idea for the `after:`. I'll take a look. The crash and the leak are weird, I'm not sure of why it behaves like this.

dunglas commented:
Maybe related: https://github.com/dunglas/frankenphp/pull/81

withinboredom commented on frankenphp.c:
```diff
@@ -105,13 +106,24 @@ static uintptr_t frankenphp_clean_server_context() {
-      zval_ptr_dtor(&PG(http_globals)[i]);
+      // Restore worker script super globals
+      for (i = 0; i < NUM_TRACK_VARS; i++) {
+        ZVAL_COPY(&PG(http_globals)[i], &ctx->worker_http_globals[i]);
```
You may have to call `GC_DEREF(&PG(http_globals)[i])` here to prevent a memory leak. I think. 

This is because you've increased the references here +1, but in `php_has_environment()` the ref will also increase +1. (IIRC) So you need to manually decrease it -1.

withinboredom replied:
or `GC_DEREF(&ctx->worker_http_global[i])`, :D

TimWolla replied:
If that's the problem, `ZVAL_COPY_VALUE` will copy the value, without increasing the reference count. But then `memcpy` should also have worked.

withinboredom replied:
I believe it may be only needed in this if-branch.

dunglas replied:
I "fixed" the leak but as you can see the fix is bad, there is a logic error somewhere :D

dunglas commented:
@TimWolla @withinboredom would you mind testing/reviewing thi patch one more time? The current implementation looks ok-ish to me.

TimWolla commented:
@dunglas For the test script in https://github.com/dunglas/frankenphp/pull/796#pullrequestreview-2067722841 I'm still seeing the wrong data being written to the `log` file:

```
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
before: worker-getopt.php 
inside: /worker-getopt.php?some_query_parameters 127.0.0.1
after : /worker-getopt.php?some_query_parameters 127.0.0.1
before: /worker-getopt.php?some_query_parameters 127.0.0.1
```

dunglas commented on frankenphp.c:
```diff
@@ -105,13 +106,23 @@ static uintptr_t frankenphp_clean_server_context() {
+      for (i = 0; i < NUM_TRACK_VARS; i++) {
+        zval_ptr_dtor(&PG(http_globals)[i]);
+      }
+      memcpy(&PG(http_globals), &ctx->worker_http_globals,
```
This doesn't work because `php_hash_environment()` clears `PG(http_globals)`): https://github.com/php/php-src/blob/62cdaac2806ba0a9f3443508ba60a6f552ef6bce/main/php_variables.c#L787

dunglas commented on frankenphp.c:
```diff
@@ -105,13 +106,23 @@ static uintptr_t frankenphp_clean_server_context() {
 
-    for (i = 0; i < NUM_TRACK_VARS; i++) {
-      zval_ptr_dtor(&PG(http_globals)[i]);
+      php_hash_environment();
```
For this to work, we need to save and restore the `SG(request_info)` of the worker script, but I wonder if it is worth it. Resetting the whole context as in the beginning of the script will be costly and useless. After all, maybe should we keep the current behavior and keep the context of the previous request until the next one is handled.

WDYT @withinboredom @TimWolla?

TimWolla replied:
As I'm mentioned in the issue I created, I don’t have an opinion here. That depends on what you believe your users expectations are.

withinboredom replied:
That's a tricky one. Personally, I use the last context to run logging things (treblle specifically) before continuing to the next request so I would find it useful. I can't think of why you would not want the last request's context.

dunglas commented:
In the last commit, I fixed the crash without changing the existing behavior (the docs change is now irrelevant, I'll revert that). I think it's good enough for now.

TimWolla commented:
> In the last commit, I fixed the crash without changing the existing behavior (the docs change is now irrelevant, I'll revert that). I think it's good enough for now.

I can confirm the crash is fixed, but internal functions see an empty `$_SERVER` now. That's certainly better than a crash, but would still be inconsistent with userland’s PoV.

dunglas commented on frankenphp.c:
```diff
@@ -73,11 +73,11 @@ typedef struct frankenphp_server_context {
-    return 0;
-  }
+
+  free(ctx->cookie_data);
```
Moving this `free()` here likely fix a memory leak in worker mode.

dunglas commented:
I cleaned the code and updated the docs. I think I've found a way to keep the userland and internal structures in sync but I didn't create a test case. @TimWolla would you mind checking if this now works with the Tideways extension?

TimWolla commented:
> I think I've found a way to keep the userland and internal structures in sync but I didn't create a test case.

I can confirm that internal functions called after `frankenphp_handle_request()` see the same `$_SERVER` as userland *initially* saw.

However what is not kept in sync is manually writing to `$_SERVER` either within the callback or after `frankenphp_handle_request()` finishes and before calling the internal function.

```php
$_SERVER['dummy'] = 'value';
```

Dumping `var_dump(count($_SERVER))` returns `42` in my test case, but an internal function
 
```c
ZEND_FUNCTION(native_debug)
{
    zend_is_auto_global(ZSTR_KNOWN(ZEND_STR_AUTOGLOBAL_SERVER));
    HashTable *server = Z_ARRVAL_P(&PG(http_globals)[TRACK_VARS_SERVER]);
    printf("%d\n", zend_hash_num_elements(server));
    zval *request_uri = zend_hash_str_find(server, "REQUEST_URI", strlen("REQUEST_URI"));
    printf("%s\n", Z_STRVAL_P(request_uri));
}
```
prints 41 (because `dummy` is missing). Likewise modifying an entry using:

```php
$_SERVER['REQUEST_URI'] = 'dummy';
```

is not visible to the native function.

---------------

I'm not sure if as a user I would prefer a crash or silently incorrect data. Your call.

dunglas commented:
@TimWolla are the two data structures synced automatically in a normal script? I may be wrong, but I don't think that we do anything special FrankenPHP-side that would cause this desync.

dunglas commented:
@TimWolla according to my tests, this behaves the same even when not in worker mode, so it looks "expected" to me, at least let's say that it's not related to FrankenPHP.

TimWolla commented:
@dunglas Can confirm with the cli-server. Sorry for the confusion. This looks good to me then!
